### PR TITLE
feat(API): add /diagnostics endpoint for system-wide debug snapshot

### PR DESCRIPTION
**Pull request by TheLastCicada**
## Summary

Adds a new top-level `GET /diagnostics` endpoint that returns a single JSON object summarizing CADT, Chia, DataLayer, and host machine state — intended for sysadmins and users debugging a CADT install. Mounted as a sibling of `/health` (not under `/v1` or `/v2`); branched off `v2-rc2`.

### What the response includes

- **CADT**: version, config directory + file, V1/V2 database paths, datalayer URL + file-server URL, simulator flag, per-version `{ enabled, readOnly, isGovernanceBody, apiKeyConfigured, governanceBodyId, homeOrgId }`.
- **Chia network**: configured (from CADT config) vs actual (from wallet `get_network_info`) and a `matches` flag.
- **Wallet**: `rpcUrl`, `reachable`, `connectionError` (string when unreachable, including for `AggregateError`s with empty messages), `synced`, `balanceXch`, pending-transactions block (reuses `wallet-health.js` shaping), trusted-peer cross-reference against `wallet.trusted_peers` from chia's `config.yaml`.
- **Full node**: best-effort mTLS call to local full-node RPC (`get_blockchain_state` + `get_connections`) — degrades to `reachable: false` if the certs aren't present.
- **DataLayer**: reachability + subscriptions list with per-store generation/target-generation and a `synced` flag (`Number.isFinite` guard so `undefined === undefined` doesn't falsely report synced); bounded-concurrency worker pool with a 30 s wall-clock budget that emits `truncated: true` if cut off.
- **Services flags**: aggregate `walletReachable` / `fullNodeReachable` / `datalayerReachable`.
- **chia-tools probe**: `chia-tools --version` (then `version` as fallback) with a 5 s timeout; distinguishes "not installed" from "installed but broken".
- **Process scan**: POSIX `ps -eo` scan for chia binaries with multi-version detection; reports `supported: false` on Windows.
- **System**: platform/arch, CPU model + cores (via `os.cpus()`), total/free RAM, free/total disk on the partition holding `${CHIA_ROOT}` (via `fs.promises.statfs`).

### Design properties

- **Degrades gracefully**: every external call is wrapped in `settle(label, producer, timeoutMs)` that races against a hard wall-clock deadline. Failures become `{ ok: false, error }` rather than throwing; one wedged subsystem can't block the rest of the response. Worst-case wall-clock ~30 s; healthy responses come back in well under a second.
- **Survives "the wallet is down" / "migrations are slow"**: `/diagnostics` lives in `HEALTH_ENDPOINTS`, and `isHealthEndpoint` skips were added to the wallet-synced, home-org-synced, and all-data-synced header middlewares so the endpoint isn't blocked by the wallet RPC's 300 s socket timeout or by `waitForMigrations()` when the DB layer itself is the broken subsystem.
- **Auth**: enforced by the existing global API-key middleware. No duplicate check needed.
- **READ_ONLY**: response is reduced to non-sensitive public fields (CADT version, configured network, system info, chia-tools probe, process scan) and short-circuits **before** the wallet/datalayer/full-node RPC fan-out — matches the precedent in [`src/routes/wallet-health.js`](src/routes/wallet-health.js) so a public observer node doesn't make per-request authenticated wallet RPC calls on unauthenticated public hits.

## Files

| File | Purpose |
|---|---|
| `src/routes/diagnostics.js` | Response builder, `settle`, `collectSubscriptions`, `buildTrustedPeerView`, `normalizeNodeId` |
| `src/utils/system-info.js` | CPU/RAM (`os`) + disk (`fs.promises.statfs`) |
| `src/utils/chia-process-scan.js` | POSIX `ps` scan + chia-binary regex |
| `src/utils/chia-tools-probe.js` | `chia-tools --version` / `version` fallback, `extractVersion` regex |
| `src/datalayer/fullNodeRpc.js` | mTLS client for `get_blockchain_state` / `get_connections` |
| `src/datalayer/wallet.js` | adds `getWalletConnections()` for trusted-peer cross-reference |
| `src/middleware.js` | mounts `GET /diagnostics`, adds it to `HEALTH_ENDPOINTS`, adds health-endpoint skip to three header middlewares |
| `tests/v2/integration/diagnostics.spec.js` | HTTP integration tests (22 cases) |
| `tests/v2/integration/diagnostics-helpers.spec.js` | Pure unit tests for the helpers (24 cases) |
| `tests/v2/live-api/wallet-health.live.spec.js` | Live-API smoke test against a real CADT install |

## Test plan

- [x] `npm run test:v1` — 146 passing, 0 failing
- [x] `npm run test:v2` — 1619 passing, 0 failing (includes 46 new diagnostics tests)
- [x] Pre-push multi-persona review (10 reviewers in parallel: senior-engineer, simplicity/reuse, scope/tone/correctness, domain-context, root-cause/layer, concurrency/cancellation, protocol/schema, observability/diagnostics, test-quality/coverage, generic pre-push-diff-review). All critical findings addressed; warnings/info items either fixed or deliberately deferred with rationale.
- [ ] Live-API smoke check via `npm run test:v2:live:wallet-health` against a real CADT + Chia install (runs in CI; the live test added a `/diagnostics` describe block that asserts response shape, types, and reasonable value ranges without pinning specific environment values).

### Matrix coverage (READ_ONLY × API key configured × key provided)

| READ_ONLY | API key | Key in request | Expected | Test |
|---|---|---|---|---|
| false | no | n/a | 200 full | basic-shape tests |
| false | yes | correct | 200 full | `accepts requests with the correct x-api-key header` |
| false | yes | wrong (length mismatch) | 403 | `rejects requests with a wrong-length x-api-key header` |
| false | yes | wrong (equal length) | 403 | `rejects requests with a wrong-but-equal-length x-api-key header` |
| false | yes | none | 403 | `rejects unauthenticated requests when CADT_API_KEY is configured` |
| true | no | n/a | 200 reduced | `public observer node (READ_ONLY=true, no API key): returns 200 with reduced fields without auth` |
| true | yes | correct | 200 reduced | `READ_ONLY=true + API key configured + correct key: returns 200 with reduced fields` |
| true | yes | wrong | 403 | `READ_ONLY=true + API key configured + wrong key: rejects with 403` |
| true | yes | none | 403 | `READ_ONLY=true + API key configured + no key provided: rejects with 403` |

## Manual smoke check (suggested for reviewers)

```bash
# Healthy install
curl -s -H "x-api-key: \$CADT_API_KEY" http://localhost:31310/diagnostics | jq .

# With wallet stopped (verify graceful degradation)
chia stop wallet
curl -s -H "x-api-key: \$CADT_API_KEY" http://localhost:31310/diagnostics | jq '.chia.wallet'
# Expect: { reachable: false, connectionError: "Wallet RPC at ... did not respond", ... }
chia start wallet
```

---

> [!NOTE]
> **Medium Risk**
> Adds a new unauthenticated-by-default (unless API key configured) diagnostics surface that executes local process/CLI probes and multiple RPC calls; mistakes could leak sensitive operational data or add startup/runtime latency despite timeouts and read-only gating.
> 
> **Overview**
> Adds a new top-level `GET /diagnostics` endpoint that returns a single JSON snapshot of CADT config/state plus Chia wallet/full-node/datalayer reachability and host system metrics, with per-section `ok/warning/critical` status aggregation and hard timeouts to degrade gracefully.
> 
> To support this, introduces new helpers for OS info (`system-info.js`), local Chia process scanning (`chia-process-scan.js`), `chia-tools` detection (`chia-tools-probe.js`), and a minimal mTLS full-node RPC client (`fullNodeRpc.js`), plus extends `wallet.js` with `get_version` and `get_connections` helpers.
> 
> Middleware now treats `/diagnostics` as a health-style endpoint (bypassing rate limits, startup gates, and synced-header probes) while explicitly blocking the route when `READ_ONLY` is enabled; startup additionally logs a fire-and-forget diagnostics snapshot after migrations complete. Tests add broad unit/integration/live-api coverage for response shape, auth gating, and helper edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 28b0a1eaaa4473f744159406b1b82063a48ffd6f. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/routes/diagnostics.js [hunk 1]
```diff
@@ -0,0 +1,541 @@
...
+  // make /diagnostics itself fail whenever the DB/wallet modules can't
+  // initialize -- which is precisely the failure mode this endpoint exists
+  // to diagnose.
+  if (readOnly) {
+    return buildReadOnlyResponse({ timestamp, configV1, configV2, appConfig });
+  }
+
+  const wallet = (await import('../datalayer/wallet.js')).default;
+  const fullNodeRpc = (await import('../datalayer/fullNodeRpc.js')).default;
+  const persistance = await import('../datalayer/persistance.js');
```
[the remark this answers is not in the document]
Comment (TheLastCicada) [reply src/routes/diagnostics.js:253]: not valid finding

File: src/routes/diagnostics.js [hunk 2]
```diff
@@ -0,0 +1,550 @@
...
+      if (Date.now() - startedAt >= SUBSCRIPTION_BUDGET_MS) {
+        truncated = true;
+        return;
+      }
+      await processOne(storeIds[i], i);
+    }
+  };
+
+  await Promise.all(
+    Array.from({ length: Math.min(SUBSCRIPTION_CONCURRENCY, storeIds.length) }, () => worker()),
```
[the remark this answers is not in the document]
Comment (TheLastCicada) [reply src/routes/diagnostics.js:166]: False positive — the worker pool is correct for single-threaded JS:

1. **`nextIndex++` is atomic in the event loop.** Each worker yields at `await processOne(...)`, and by that point the index is already claimed and bounds-checked. There's no interleaving of the increment and the guard.

2. **Every path that skips `results[index]` sets `truncated = true`.** The only way a gap appears in the sparse array is when the budget check fires (line 97 in `processOne` or line 132 in the worker loop), and both set `truncated = true` before returning. So `subscriptions.length < totalSubscriptions` always coincides with `truncated === true`.

3. **Errors never leave gaps.** The `catch` on line 122 always populates `results[index]`, and the `!syncStatus` branch on line 105 does too.

Resolving as not actionable.

File: src/middleware.js [hunk 3]
```diff
@@ -490,6 +513,38 @@ app.get('/health', (req, res) => {
...
+// check here.
+//
+// Read-only: when V1 or V2 READ_ONLY is set we strip sensitive fields
+// (balances, transaction details, peer details, subscription IDs, home org
+// IDs) the same way wallet-health.js does for /v{1,2}/health/wallet.
+app.get('/diagnostics', async (req, res) => {
+  try {
+    const configV1 = getConfig();
+    const configV2 = getConfigV2();
+    const readOnly = configV2.READ_ONLY === true || configV1.READ_ONLY === true;
```
[the remark this answers is not in the document]
Comment (TheLastCicada) [reply src/middleware.js]: Valid finding. The diagnostics handler used `=== true` while the rest of middleware uses truthy checks (`|| false`). Fixed to `!!(configV2.READ_ONLY || configV1.READ_ONLY)` to match.

File: src/routes/diagnostics.js [hunk 4]
```diff
@@ -0,0 +1,562 @@
...
+    : { matches: [], installPaths: [], multipleVersionsDetected: false, error: chiaProcessesRes.error };
+
+  // If the scan succeeded and found processes, check for chia_full_node.
+  // If the scan failed or is unsupported (Windows, minimal Docker image
+  // without ps, /proc restrictions), we can't tell whether the full node
+  // is running, so default to true and let the RPC timeout be the backstop.
+  const scanReliable = chiaProcessesRes.ok && !processesValue.note;
+  const fullNodeRunningLocally = scanReliable
+    ? processesValue.matches.some((m) => /chia_full_node/i.test(m.command))
+    : true;
```
[the remark this answers is not in the document]
Comment (TheLastCicada) [reply src/routes/diagnostics.js:278]: True positive — good catch. The `scanReliable` guard only checked for `processesValue.note` (the Windows early-return path) but missed `processesValue.error` (the POSIX catch block when `ps` fails). When `ps` fails in a container/restricted environment, the function returns `{ matches: [], error: "..." }` without a `note`, so `scanReliable` was incorrectly `true`, causing `fullNodeRunningLocally` to be `false` and skipping RPCs.

Fixed by extending the check to `chiaProcessesRes.ok && !processesValue.note && !processesValue.error`.

File: src/routes/diagnostics.js [hunk 5]
```diff
@@ -0,0 +1,565 @@
...
+    view.connected = connections.map((c) => {
+      const trusted = !!(
+        normalizedTrustedSet && normalizedTrustedSet.has(normalizeNodeId(c.nodeId))
+      );
+      if (trusted) view.hasTrustedConnection = true;
+      return { peerHost: c.peerHost, peerPort: c.peerPort, type: c.type, trusted };
+    });
+  } else if (connectionsResult) {
+    view.connectionsError = connectionsResult.error;
+  }
```
[the remark this answers is not in the document]
Comment (TheLastCicada) [reply src/routes/diagnostics.js:232]: True positive. `getWalletConnections` catches errors internally and returns `{ success: false, error }` instead of throwing, so `settle()` always wraps it as `{ ok: true, value: { success: false, error } }`. The `else if` branch that sets `connectionsError` was unreachable for wallet failures, and `connected` silently became `[]` with no error indication.

Fixed by also checking `connectionsResult.value?.success !== false` and surfacing the inner error when `success` is `false`.

File: src/datalayer/wallet.js [hunk 6]
```diff
@@ -329,6 +329,67 @@ const getActiveNetwork = async () => {
...
+    .cert(cert)
+    .timeout(timeout)
+    .send(JSON.stringify({}));
+
+  const data = response.body;
+  if (data.success) {
+    return data.version || null;
+  }
+  return null;
+};
```
[the remark this answers is not in the document]
Comment (TheLastCicada) [reply src/datalayer/wallet.js:353]: True positive. Added try-catch with `logger.debug()` returning `null`, consistent with sibling RPC functions. This makes the function safe for any future caller outside of `settle()`.

File: src/datalayer/fullNodeRpc.js [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+ * Like the wallet client, we use mTLS with the standard chia SSL files
+ * under `${chiaRoot}/config/ssl/full_node/`. If those files are missing
+ * (e.g. on a CADT-only host), the helpers return `{ reachable: false }`
+ * instead of throwing.
+ */
+
+import fs from 'fs';
+import path from 'path';
+import superagent from 'superagent';
+
```
[the remark this answers is not in the document]
Comment (TheLastCicada) [reply src/datalayer/fullNodeRpc.js:16]: True positive. `fullNodeRpc.js` is a new file in this PR and was relying on `wallet.js` being imported first as a side effect. Added explicit `process.env["NODE_TLS_REJECT_UNAUTHORIZED"] = 0` to match `wallet.js` and `persistance.js`.

File: src/routes/diagnostics.js [hunk 8]
```diff
@@ -0,0 +1,485 @@
...
+    settle('wallet.getActiveNetwork', () => wallet.getActiveNetwork(), DEFAULT_TIMEOUT_MS),
+    settle('wallet.getChiaVersion', () => wallet.getChiaVersion(), DEFAULT_TIMEOUT_MS),
+    settle('wallet.walletIsSynced', () => wallet.walletIsSynced(), DEFAULT_TIMEOUT_MS),
+    settle('wallet.getWalletBalance', () => wallet.getWalletBalance(), DEFAULT_TIMEOUT_MS),
+    settle('wallet.getWalletConnections', () => wallet.getWalletConnections(), DEFAULT_TIMEOUT_MS),
+    settle(
+      'getWalletHealthResponse',
+      () => getWalletHealthResponse(wallet, { readOnly: false }),
+      DEFAULT_TIMEOUT_MS,
+    ),
```
[the remark this answers is not in the document]
Comment (TheLastCicada) [reply src/routes/diagnostics.js:292]: True positive. Removed the direct `walletIsSynced` settle call — `getWalletHealthResponse` already calls it internally, so the duplicate created a race on `lastWalletSyncError` (each call clears it to `null` at the top). Now `synced` is derived from `walletHealthRes.value?.synced` and the error slot is only written by a single caller.

File: src/routes/index.js [hunk 9]
```diff
@@ -73,6 +74,13 @@ export const initializeDatabases = async () => {
     migrationsReady = true;
     logger.info('All database migrations completed');
 
+    // Snapshot diagnostics into the log so operators always have a baseline,
+    // even on READ_ONLY nodes that can't serve the /diagnostics endpoint.
+    // Fire-and-forget: don't block scheduler start on RPC timeouts.
+    getDiagnosticsResponse()
+      .then((d) => logger.info(`Startup diagnostics: ${JSON.stringify(d)}`))
+      .catch((e) => logger.warn(`Startup diagnostics collection failed: ${e.message}`));
```
[the remark this answers is not in the document]
Comment (TheLastCicada) [reply src/routes/index.js:82]: The whole point of logging it at startup was specifically so READ_ONLY nodes also get a diagnostics baseline somewhere, just not exposed over the network. The log file is not a public endpoint.
